csrc/preseg_passes/move_pad.cpp

```diff
@@ -0,0 +1,530 @@
+  return true;
+}
+
+// check if `use` and `p` are the same pad operatoin.
+bool isSamePadOp(Expr* use, PadOp* p) {
+  if (!use->isA<PadOp>()) {
+    return false;
+  }
+
+  auto* use_pad = use->as<PadOp>();
+  if (!simplifyExpr(SimplifyingIrBuilder::eqExpr(use_pad->value(), p->value()))
+           ->isTrue()) {
+    return false;
+  }
+
+  if (use_pad->getPadWidths().size() != p->getPadWidths().size()) {
+    return false;
+  }
+
+  std::vector<int64_t> padded_axes = p->getPaddedAxes();
+  if (use_pad->getPaddedAxes() != padded_axes) {
+    return false;
+  }
+
+  for (auto idx : padded_axes) {
+    if (!simplifyExpr(
+             SimplifyingIrBuilder::eqExpr(
+                 use_pad->getPadWidths(idx).first, p->getPadWidths(idx).first))
+             ->isTrue() ||
+        !simplifyExpr(SimplifyingIrBuilder::eqExpr(
+                          use_pad->getPadWidths(idx).second,
+                          p->getPadWidths(idx).second))
+             ->isTrue()) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+// This operation replaces:
+//   CatOp(inputs)
+// with:
+//   BinaryOp(inputs[n-1], BinaryOp(inputs[n-2], BinaryOp(..., BinaryOp[0])...))
+Val* replayCatOpWithBinaryOp(
+    const std::vector<Val*>& inputs,
+    DataType data_type) {
+  // replay `CatOp` with series of BinaryOp instead, since we might have
+  // pushed `PadOp` out and breaking the codegen if `CatOp` remains.
+  Val* res = nullptr;
+  bool is_boolean = isBooleanType(data_type);
+  for (Val* inp : inputs) {
+    if (res == nullptr) {
+      res = inp;
+    } else {
+      if (is_boolean) {
+        res = bitwise_or(res, inp);
+      } else {
+        res = add(res, inp);
+      }
+    }
+  }
+  // restore data type if it's promoted by BinaryOp.
+  return maybeCastOp(data_type, res);
+}
+
+// The pass assumes propagating PadOp with zero pad. The criteria here for
+// return true is that `unaryOp(0) == 0`
+bool zeroIsFixedPoint(UnaryOpType t) {
+  switch (t) {
+    case UnaryOpType::Cast:
+    case UnaryOpType::Abs:
+    case UnaryOpType::Asin:
+    case UnaryOpType::Asinh:
+    case UnaryOpType::Atan:
+    case UnaryOpType::Atanh:
+    case UnaryOpType::Ceil:
+    case UnaryOpType::Erf:
+    case UnaryOpType::Erfinv:
+    case UnaryOpType::Floor:
+    case UnaryOpType::Gelu:
+    case UnaryOpType::Imag:
+    case UnaryOpType::Silu:
+    case UnaryOpType::Neg:
+    case UnaryOpType::Real:
+    case UnaryOpType::Relu:
+    case UnaryOpType::Round:
+    case UnaryOpType::Sin:
+    case UnaryOpType::Sinh:
+    case UnaryOpType::Sqrt:
+    case UnaryOpType::Tan:
+    case UnaryOpType::Tanh:
+    case UnaryOpType::Trunc:
+      return true;
+    default:
+      return false;
+  }
+}
+
+// The pass assumes propagating PadOp with zero pad. The criteria here for
+// return true is that `binaryOp(0, x) == 0` & `binaryOp(x, 0) == 0`
+bool zeroIsIdentity(BinaryOpType t) {
+  switch (t) {
+    case BinaryOpType::Add:
+    case BinaryOpType::Mul:
+    case BinaryOpType::Sub:
```

Comment:
This does not hold for `Sub`: `0 - x != x`.

Reply:
ohohoh good catch again. I need better testing for these....

Reply:
:eyes: